### ISSUE TITLE
Seperate steps in `run_cargo_tests`

### ISSUE THREAD

**Issue by mariaschett** [project-oak/oak#2454]
Seperate the three steps from `run_cargo_tests`:

- `run_cargo_clippy`
- `run_cargo_test`
- `run_cargo_doc`

by adding a two new flags to (i) run cargo clippy and (ii) run cargo doc.

Then, in `run-ci` the three steps can run in parallel, which should reduce the overall time (as we cache the builds now #2436):

Some data points for #2444 :
- `run_cargo_clippy` took ~ 19 min
- `run_cargo_test` took ~ 26 min
- `run_cargo_doc` ~ < 2 min

Seperating `run_cargo_doc` is not necessary for performance but improves the interface: the `run_cargo_test` command then only runs the cargo tests.

**Comment by tiziano88** [comment]
The downside of separating the steps is that they use up more machines from the available pool, so it may actually result in degraded performance, especially if there are concurrent PRs (since they would be queued for execution instead of starting right away). I don't actually know what the pool size (i.e. available parallelism) is though.

**Comment by mariaschett** [comment]
Good point, but this is currently our longest running task. 

I think we would still want to split this one, and think about merging shorter ones to reduce the machines required from the pool.

Either way, I'll try to find out how many those are.

**Comment by mariaschett** [comment]
It seems we can have [20 concurrent jobs](https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits)---so we [can have 3 more](https://github.com/project-oak/oak/blob/9d62a035906a098a4a2d694babcbd836902be01c/.github/workflows/ci.yaml#L25). :)

However, it might me a good idea to check if we can group jobs together---especially as IIUC we have a couple of minutes startup time per job.

maybe relevant @conradgrobler , too?

**Comment by tiziano88** [comment]
That means that if two people have a PR out for review, or if another PR was just submitted, the later PR will have to wait a long time until any of its jobs even gets scheduled. Given that, I'm not convinced we should split things more. Or do you think this would not happen in practice?

**Comment by mariaschett** [comment]
True, we don't want to exhaust the 20 concurrent jobs.

Nevertheless I think it is best to reduce the longest running job---which `run-cargo-tests` is. 

If we want to reduce concurrent jobs, I think it would be better to combine tasks, which finish faster. To give an example, run-cargo-deny finishes in 9 min, check-format in 11min, so if we combine those we safe jobs but don't add to the overall time the PR needs to finish).

I optimistically went ahead and seperated `run-cargo-clippy` and `run-cargo-tests` in the linked PR  #2458 .

It does seem to help a bit: when seperated, `run-cargo-tests` runs in 54 min, `run-cargo-clippy` in 24m in #2458 . This is a 20 min reduction from a total of 1h 15 m for the combined `run-cargo-tests` from a similar PR #2453  earlier today. 

However, the total time a job needs seems to fluctuate a lot from PR to PR. Looking just at those two PRs #2458  and #2453 for some jobs (e.g., check-format, run-cargo-deny) we have differences of taking nearly double time.

Do you have any idea why this is the case? 

The length a job takes should not be affected by the resources available, or? It should be just it's starting time, which might be later. Or is my mental model wrong here?

**Comment by mariaschett** [comment]
I suggest to keep `run_cargo_doc` as part of `run_cargo_tests` (to not create too many concurrent small jobs), but split the run-cargo-clippy in a seperate task, because it does seem to save some overall time.

@tiziano88 do you agree?

**Comment by tiziano88** [comment]
Yeah LGTM, we can try it out and see how it goes, thanks!
